crates/client/src/keyring/mod.rs

```diff
@@ -0,0 +1,312 @@
+    /// The default backend when no configuration option is set
+    pub const DEFAULT_BACKEND: &'static str = Self::SUPPORTED_BACKENDS[0];
+
+    /// Returns a human-readable description of a keyring backend.
+    pub fn describe_backend(backend: &str) -> &'static str {
+        match backend {
+            "secret-service" => "Freedesktop.org secret service (GNOME Keyring or KWallet)",
+            "linux-keyutils" => "Linux kernel memory-based keystore (lacks persistence, not suitable for desktop use)",
+            "windows" => "Windows Credential Manager",
+            "macos" => "MacOS Keychain",
+            "ios" => "Apple iOS Keychain",
+            "mock" => "Mock credential store with no persistence (for testing only)",
+            _ => "(no description available)"
+        }
+    }
+
+    fn load_backend(backend: &str) -> Option<Box<keyring::CredentialBuilder>> {
+        if !Self::SUPPORTED_BACKENDS.contains(&backend) {
+            return None;
+        }
+
+        #[cfg(any(target_os = "linux", target_os = "freebsd", target_os = "openbsd"))]
+        if backend == "secret-service" {
+            return Some(keyring::secret_service::default_credential_builder());
+        }
+
+        #[cfg(target_os = "linux")]
+        if backend == "linux-keyutils" {
+            return Some(keyring::keyutils::default_credential_builder());
+        }
+
+        #[cfg(target_os = "macos")]
+        if backend == "macos" {
+            return Some(keyring::macos::default_credential_builder());
+        }
+
+        #[cfg(target_os = "ios")]
+        if backend == "ios" {
+            return Some(keyring::ios::default_credential_builder());
+        }
+
+        #[cfg(target_os = "windows")]
+        if backend == "windows" {
+            return Some(keyring::windows::default_credential_builder());
+        }
+
+        if backend == "mock" {
+            return Some(keyring::mock::default_credential_builder());
+        }
+
+        unreachable!("missing logic for backend {backend}")
+    }
+
+    /// Instantiate a new keyring.
+    ///
+    /// The argument should be an element of [Self::SUPPORTED_BACKENDS].
+    pub fn new(backend: &str) -> Result<Self> {
+        Self::load_backend(backend)
+            .ok_or_else(|| KeyringError::unknown_backend(backend.to_string()))
+            .map(|imp| Self {
+                imp,
+                // Get an equivalent &'static str from our &str
+                name: Self::SUPPORTED_BACKENDS
+                    .iter()
+                    .find(|s| **s == backend)
+                    .expect("successfully-loaded backend should be found in SUPPORTED_BACKENDS"),
+            })
+    }
+
+    /// Instantiate a new keyring using the backend specified in a configuration file.
+    pub fn from_config(config: &Config) -> Result<Self> {
+        if let Some(ref backend) = config.keyring_backend {
+            Self::new(backend.as_str())
+        } else {
+            Self::new(Self::SUPPORTED_BACKENDS[0])
```

Comment:
```suggestion
            Self::new(Self::DEFAULT_BACKEND)
```

Reply:
Calvin already merged, so I'll fix this in my next PR.